File: Source/Java/Expressive/src/main/java/com/bijington/expressive/functions/statistical/MeanFunction.java
package com.bijington.expressive.functions.statistical;

import com.bijington.expressive.exceptions.ExpressiveException;
import com.bijington.expressive.expressions.IExpression;
import com.bijington.expressive.functions.FunctionBase;
import com.bijington.expressive.helpers.Convert;
import com.bijington.expressive.helpers.Numbers;

import java.lang.reflect.Array;

/**
 * Created by shaun on 18/07/2016.
 */
public class MeanFunction extends FunctionBase {
    @Override
    public String getName() {
        return "Mean";
    }

    @Override
    public Object evaluate(IExpression[] parameters) throws ExpressiveException {
        this.validateParameterCount(parameters, -1, 1);

        int count = 0;
        Object result = 0;

        for (int parameterIndex = 0; parameterIndex < parameters.length; parameterIndex++) {
            IExpression value = parameters[parameterIndex];

            int increment = 1;
            Object evaluatedValue = value.evaluate(this.getVariables());

            if (evaluatedValue != null &&
                evaluatedValue.getClass().isArray()) {
                int enumerableCount = 0;
                Object enumerableSum = 0;

                for (int j = 0; j < Array.getLength(evaluatedValue); j++) {
                    Object arrayValue = Array.get(evaluatedValue, j);

                    enumerableCount++;
                    enumerableSum = Numbers.add(enumerableSum, arrayValue);
                }

                increment = enumerableCount;
                evaluatedValue = enumerableSum;
            }

            result = Numbers.add(result, evaluatedValue);
            count += increment;
        }

        return Convert.toDouble(result) / count;
    }
}

File: Source/Expressive/Functions/Statistical/MeanFunction.cs
using Expressive.Expressions;
using Expressive.Helpers;
using System;
using System.Collections;
using System.Collections.Generic;

namespace Expressive.Functions.Statistical
{
    internal class MeanFunction : FunctionBase
    {
        public override string Name => "Mean";

        public override object Evaluate(IExpression[] parameters, Context context)
        {
            this.ValidateParameterCount(parameters, -1, 1);

            return Evaluate(parameters, Variables);
        }

        internal static object Evaluate(IExpression[] parameters, IDictionary<string, object> variables)
        {
            var count = 0;
            object result = 0;

            foreach (var value in parameters)
            {
                var increment = 1;
                var evaluatedValue = value.Evaluate(variables);

                if (evaluatedValue is IEnumerable enumerable)
                {
                    var enumerableCount = 0;
                    object enumerableSum = 0;

                    foreach (var item in enumerable)
                    {
                        if (item is null)
                        {
                            continue;
                        }

                        enumerableCount++;
                        enumerableSum = Numbers.Add(enumerableSum, item);
                    }

                    increment = enumerableCount;
                    evaluatedValue = enumerableSum;
                }
                else if (evaluatedValue is null)
                {
                    continue;
                }

                result = Numbers.Add(result, evaluatedValue);
                count += increment;
            }

            return Convert.ToDouble(result) / count;
        }
    }
}
